doc/lspconfig.txt

```diff
@@ -1,242 +1,628 @@
+or special keys like `description`. Example:
 
+>
+  commands = {
+    TexlabBuild = {
+      function()
+        buf_build(0)
+      end;
+      "-range";
+      description = "Build the current buffer";
+    };
+  };
+<
 
-Attempt to run the language server, and open the log with:
+The `configs.__newindex` metamethod consumes the config definition and returns
+an object with a `setup()` method, to be invoked by users:
+>
+    require'lspconfig'.SERVER_NAME.setup{}
+
+After you set `configs.SERVER_NAME` you can add arbitrary language-specific
+functions to it if necessary.
+
+Example:
 
 >
-    :lua vim.cmd('e'..vim.lsp.get_log_path())
+    configs.texlab.buf_build = buf_build
 <
+==============================================================================
+ADDING NEW SERVERS                                    *lspconfig-adding-servers*
+
+The three steps for adding and enabling a a new server configuration are:
 
+- load the lspconfig module (note, this is a stylistic choice)
+>
+   local lspconfig = require 'lspconfig'
+<
+- define the configuration
 
-Most of the time, the reason for failure is present in the logs.
+>
+   local configs = require 'lspconfig/configs'
+
+   -- Check if the config is already defined (useful when reloading this file)
+   if not lspconfig.foo_lsp then
+     configs.foo_lsp = {
+       default_config = {
+         cmd = {'/home/neovim/lua-language-server/run.sh'};
+         filetypes = {'lua'};
+         root_dir = function(fname)
+           return lspconfig.util.find_git_ancestor(fname)
+         end;
+         settings = {};
+       };
+     }
+   end
+
+- call `setup()` to enable the FileType autocmd
+>
+   lspconfig.foo_lsp.setup{}
+<
 ==============================================================================
-9. Built-in commands                             *lspconfig-built-in-commands*
+INDEX                                                          *lspconfig-index*
 
+The autogenerated documentation for all language servers is accessible in
+`server_configurations.md`. Type `gf` on the filename from this document to jump
+to the markdown
+file.
 
-* `:LspInfo` shows the status of active and configured language servers.
-The following support tab-completion for all arguments:
-* `:LspStart <config_name>` Start the requested server name. Will only successfully start if the command detects a root directory matching the current config. Pass `autostart = false` to your `.setup{}` call for a language server if you would like to launch clients solely with this command. Defaults to all servers matching current buffer filetype.
-* `:LspStop <client_id>` Defaults to stopping all buffer clients.
-* `:LspRestart <client_id>` Defaults to restarting all buffer clients.
 ==============================================================================
-10. The wiki                                              *lspconfig-the-wiki*
+ROOT DETECTION                                        *lspconfig-root-detection*
```

Comment:
I think this section (and many users in general) would benefit from having a clear statement here at the beginning on the dual role of root directories for lspconfig:

1. the directory to pass to **servers** as `rootDirectory` (to do with whatever they do; this can include ignoring)
2. the directory for **lspconfig** to consider under the regime of a single server (i.e., anything under `root_directory` is attached to the same server instead of starting a new one).

This _is_ later explained under "single file support", but I believe it would be better to move that whole section to the beginning and treat it as an explanation of what "root directory" means in the context of lsp(config).